File: ui/src/checkers/helm_config/federation_strategy.ts
/**
 * Validates federation strategy and domain configs in brig.
 */

// Ours
import { BaseChecker, type CheckResult } from '../base_checker'
import type { DataLookup } from '../data_lookup'

export class FederationStrategyChecker extends BaseChecker {
    readonly path: string = 'helm_config/federation_strategy'
    readonly name: string = 'Federation strategy'
    readonly category: string = 'Helm / Config Validation'
    readonly interest = 'Setup' as const
    readonly explanation: string = 'Federation strategy (`allowNone`/`allowAll`/`allowDynamic`) controls which backends are allowed to federate. `allowNone` effectively disables federation even when `enableFederation` is true.'

    check(data: DataLookup): CheckResult {
        if (data.config && !data.config.options.expect_federation) {
            return { status: 'not_applicable', status_reason: 'Federation is not enabled.' }
        }

        const point = data.get('config/federation_strategy')
        if (!point?.value) return { status: 'gather_failure', status_reason: 'Federation strategy data not collected.' }

        let parsed: Record<string, unknown> | null = null
        try { parsed = JSON.parse(String(point.value)) } catch { /* ignore */ }
        if (!parsed) return { status: 'gather_failure', status_reason: 'Could not parse federation strategy data.' }

        const strategy: string = (parsed.strategy as string) ?? 'allowNone'
        const domain_configs: Array<Record<string, string>> = (parsed.domain_configs as Array<Record<string, string>>) ?? []
        const declared_domains: string[] = data.config?.options?.federation_domains ?? []

        if (strategy === 'allowNone') {
            return {
                status: 'unhealthy',
                status_reason: 'Federation strategy is `allowNone` — federation is effectively **disabled** even though `enableFederation` is true.',
                fix_hint: 'Set the strategy in brig helm values:\n```yaml\nbrig:\n  config:\n    optSettings:\n      setFederationStrategy: allowDynamic\n```',
                display_value: 'allowNone (disabled)',
                raw_output: point.raw_output,
            }
        }

        if (strategy === 'allowDynamic') {
            // Check that declared domains are in the config
            const configured_domains: string[] = domain_configs.map(c => c.domain ?? '')
            const missing: string[] = declared_domains.filter(d => d && !configured_domains.includes(d))

            if (missing.length > 0) {
                return {
                    status: 'unhealthy',
                    status_reason: `Strategy is \`allowDynamic\` but these declared federation partners are missing from \`setFederationDomainConfigs\`: **${missing.join(', ')}**.`,
                    fix_hint: missing.map(d => `Add to brig optSettings.setFederationDomainConfigs:\n  - domain: ${d}\n    search_policy: full_search`).join('\n'),
                    display_value: `allowDynamic (${missing.length} missing)`,
                    raw_output: point.raw_output,
                }
            }
        }

        return {
            status: 'healthy',
            status_reason: `Federation strategy: \`${strategy}\`${domain_configs.length > 0 ? `, ${domain_configs.length} domain(s) configured` : ''}.`,
            display_value: strategy,
            raw_output: point.raw_output,
        }
    }
}

export default FederationStrategyChecker

File: src/targets/config/federation_strategy.py
"""Reads federation strategy and domain configs from brig ConfigMap.

The federation strategy (allowNone/allowAll/allowDynamic) controls which
backends this instance will federate with. When using allowDynamic, the
setFederationDomainConfigs list must contain entries for each partner.
"""

from __future__ import annotations

# External
import json
from typing import Any

# Ours
from src.lib.base_target import BaseTarget, NotApplicableError
from src.lib.yaml_parser import parse_yaml, get_nested


class FederationStrategy(BaseTarget):
    """Read federation strategy and domain configs from brig.

    Only runs when expect_federation is true in the deployment configuration.
    """

    @property
    def description(self) -> str:
        """What this target checks."""
        return "Federation strategy and domain configuration"

    @property
    def explanation(self) -> str:
        """Why this matters."""
        return (
            "The federation strategy controls which backends are allowed to federate. "
            "'allowNone' (default) effectively disables federation even when "
            "enableFederation is true. 'allowAll' federates with any reachable backend. "
            "'allowDynamic' federates only with explicitly listed domains."
        )

    @property
    def unit(self) -> str:
        """No unit — returns structured JSON string."""
        return ""

    def collect(self) -> str:
        """Read brig ConfigMap and extract federation strategy settings.

        Returns:
            JSON string with strategy details.

        Raises:
            NotApplicableError: If federation is not expected.
            RuntimeError: If brig ConfigMap can't be fetched or parsed.
        """
        # Only run when federation is expected
        if not self.config.options.expect_federation:
            raise NotApplicableError("Federation is not enabled in the deployment configuration")

        self.terminal.step("Reading brig ConfigMap for federation strategy...")

        _result, cm_data = self.run_kubectl("configmap/brig")

        if not isinstance(cm_data, dict):
            raise RuntimeError("Cannot fetch brig ConfigMap")

        brig_yaml_str: str = cm_data.get("data", {}).get("brig.yaml", "")
        if not brig_yaml_str:
            raise RuntimeError("brig ConfigMap missing brig.yaml")

        try:
            brig_config: dict[str, Any] = parse_yaml(brig_yaml_str)
        except ValueError as e:
            raise RuntimeError(f"Cannot parse brig.yaml: {e}") from e

        # Extract federation strategy
        strategy: str = str(
            get_nested(brig_config, "optSettings.setFederationStrategy", "allowNone")
            or "allowNone"
        )

        # Extract federation domain configs (list of {domain, search_policy})
        raw_domain_configs: Any = get_nested(
            brig_config, "optSettings.setFederationDomainConfigs", []
        )
        domain_configs: list[dict[str, str]] = []
        if isinstance(raw_domain_configs, list):
            for entry in raw_domain_configs:
                if isinstance(entry, dict):
                    domain_configs.append({
                        "domain": str(entry.get("domain", "")),
                        "search_policy": str(entry.get("search_policy", "")),
                    })

        # Extract this backend's own federation domain
        federation_domain: str = str(
            get_nested(brig_config, "optSettings.setFederationDomain", "") or ""
        )

        result: dict[str, Any] = {
            "strategy": strategy,
            "domain_configs": domain_configs,
            "federation_domain": federation_domain,
            "configured_domains": [c["domain"] for c in domain_configs],
        }

        self._health_info = (
            f"Strategy: {strategy}, "
            f"domain: {federation_domain}, "
            f"partners: {len(domain_configs)}"
        )

        return json.dumps(result)
